src/particles.cpp

```diff
@@ -30,623 +10,697 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	core::matrix4 &texture_matrix;
+	const v2u32 texsize;
+	v2f framesize_ratio;
+	u32 last_time = 0;
+};
+
+class CollisionAffector : public irr::scene::IParticleAffector
 {
-	m_time += dtime;
-
-	static thread_local const float radius =
-			g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE;
-
-	bool unloaded = false;
-	bool is_attached = false;
-	v3f attached_pos = v3f(0,0,0);
-	float attached_yaw = 0;
-	if (m_attached_id != 0) {
-		if (ClientActiveObject *attached = env->getActiveObject(m_attached_id)) {
-			attached_pos = attached->getPosition() / BS;
-			attached_yaw = attached->getYaw();
-			is_attached = true;
-		} else {
-			unloaded = true;
-		}
+public:
+	CollisionAffector(ClientEnvironment *env,
+		irr::scene::IParticleSystemSceneNode *ps, bool collision_removal)
+		: env(env), ps(ps), last_time(0), collision_removal(collision_removal)
+	{
 	}
 
-	if (m_spawntime != 0) {
-		// Spawner exists for a predefined timespan
-		for (std::vector<float>::iterator i = m_spawntimes.begin();
-				i != m_spawntimes.end();) {
-			if ((*i) <= m_time && m_amount > 0) {
-				m_amount--;
+	virtual void affect(u32 now, irr::scene::SParticle *particlearray, u32 count)
+	{
+		if (last_time == 0) {
+			last_time = now;
+			return;
+		}
+		f32 dtime = (now - last_time) / 1000.0f;
+		last_time = now;
+
+		for (u32 i = 0; i < count; ++i) {
+			irr::scene::SParticle &p = particlearray[i];
+			float half_width = p.size.Width / 2;
+			aabb3f box {-half_width,-half_width,-half_width,half_width,half_width,half_width};
+			v3f camera_offset = intToFloat(env->getCameraOffset(),BS);
+
+			v3f position = p.pos + camera_offset;
+			v3f velocity = p.vector * 100.f;
 
-				// Pretend to, but don't actually spawn a particle if it is
-				// attached to an unloaded object or distant from player.
-				if (!unloaded)
-					spawnParticle(env, radius, is_attached, attached_pos, attached_yaw);
+			collisionMoveResult r = collisionMoveSimple(env, env->getGameDef(),
+					BS * 0.5, box, 0, dtime, &position, &velocity,
+					v3f {0.f, 0.f, 0.f});
 
-				i = m_spawntimes.erase(i);
+			if (collision_removal && r.collides) {
+				p.endTime = 0;
 			} else {
-				++i;
+				p.vector = velocity * 0.01f;
+				p.pos = position - camera_offset;
 			}
 		}
-	} else {
-		// Spawner exists for an infinity timespan, spawn on a per-second base
+	}
 
-		// Skip this step if attached to an unloaded object
-		if (unloaded)
-			return;
+        virtual irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const
+	{
+                return irr::scene::EPAT_NONE;
+        }
+private:
+	ClientEnvironment *env;
+	irr::scene::IParticleSystemSceneNode *const ps;
+	u32 last_time;
+	const bool collision_removal;
+};
+
+class AccelerationAffector : public irr::scene::IParticleAffector
+{
+public:
+	AccelerationAffector(const v3f &minacc, const v3f &maxacc)
+		: last_time(0)
+	{
+		disabled = minacc.X == 0 && minacc.Y == 0 && minacc.Z == 0 &&
+		maxacc.X == 0 && maxacc.Y == 0 && maxacc.Z == 0;
```

Comment:
Missing indent. Might require a threshold value to compensate calculation errors.

Reply:
I switched computers, and the indent setting was different...
How large? Like 2*eps or something?

Reply:
`minacc.getLengthSQ() < 0.001f && maxacc.getLengthSQ() < 0.001f`, for example.

Reply:
Would 2e-7 (or a bit larger, in the magnitude of [eps](https://en.wikipedia.org/wiki/Machine_epsilon) for `f32`) be ok too? I think 0.001 is too large.

Reply:
I'm not sure how often particles are accelerated by `0.0316 m/s` into one direction. You can use a lower value, such as `1E-6` or even `1E-7`. But I don't have a good feeling when going lower than that to ensure it will be detected as `disabled` when calculation errors sum up.

Reply:
Ok, I'll make it `1e-6` (an acceleration of 0.001 is still visible)